### PR TITLE
Mod system improvement Part I : Special buildings should work in  modders towns

### PR DESCRIPTION
**Pull request by ShubusCorporation**
These special buildings are ready to work in the modders towns:

 "stables"
 "brotherhoodOfSword"  (+2 morale)

 "lookoutTower"
There're 4 types of Tower have been introduced:  
- "high" (R=20), 
- "average" (R=15)
- "low" (R=10)
- "skyship" (for Grails, open entire map)

The type of lookout tower should be set by the "height" property:
"height" : "high" 

"library"
"fountainOfFortune" (+2 luck)
"mysticPond"
"manaVortex"
"portalOfSummoning"
"ballistaYard"
"freelancersGuild"
"escapeTunnel"
"artifactMerchant"
"magicUniversity"

"castleGate"
Castle Gate is a universal faction gate now.
It binds all player's towns which belong to one faction

"creatureTransformer" (only for vanila skeletons yet)
"spellPowerGarrisonBonus" (+2 sp)
"attackGarrisonBonus" (+2 attack)
"defenseGarrisonBonus" (+2 defence)

### CODE REVIEW

File: lib/CTownHandler.h [hunk 1]
```diff
@@ -77,7 +87,9 @@ class DLL_LINKAGE CBuilding
 		h & description;
 		h & requirements;
 		h & upgrade;
+		h & subId;
 		h & mode;
+		h & height;
```
Comment (alexvins) [inline lib/CTownHandler.h:92]: Missing backward compatibility handling.

File: lib/mapObjects/CGTownInstance.h [hunk 2]
```diff
@@ -97,25 +97,37 @@ class DLL_LINKAGE CGTownBuilding : public IObjectInterface
 {
 ///basic class for town structures handled as map objects
 public:
-	BuildingID ID; //from buildig list
-	si32 id; //identifies its index on towns vector
+	size_t indexOnTV; //identifies its index on towns vector
 	CGTownInstance *town;
+	CGTownBuilding() : town(nullptr) {};
```
Comment (alexvins) [inline lib/mapObjects/CGTownInstance.h:102]: IDs should also be initialized

File: lib/mapObjects/CGTownInstance.h [hunk 3]
```diff
@@ -97,25 +97,37 @@ class DLL_LINKAGE CGTownBuilding : public IObjectInterface
...
+	BuildingSubID::EBuildingSubID getBuildingSubtype() const
+	{
+		return bType;
+	}
 
 	template <typename Handler> void serialize(Handler &h, const int version)
 	{
-		h & ID;
-		h & id;
+		h & bID;
```
Comment (alexvins) [inline lib/mapObjects/CGTownInstance.h:112]: Missing backward compatibility handling.

File: lib/CTownHandler.cpp [hunk 4]
```diff
@@ -343,28 +343,93 @@ void CTownHandler::loadBuildingRequirements(CBuilding * building, const JsonNode
 	requirementsToLoad.push_back(hlp);
 }
 
+template<typename R>
+R CTownHandler::getMappedValue(const std::string key, const R defval, const std::map<std::string, R> & map) const
+{
+	for(const auto & kvp : map)
```
Comment (alexvins) [inline lib/CTownHandler.cpp:349]: You should use `map::find()` here. Also there is an almost identical function `parseByMap` in JsonNode.cpp

File: lib/CTownHandler.cpp [hunk 5]
```diff
@@ -343,28 +343,93 @@ void CTownHandler::loadBuildingRequirements(CBuilding * building, const JsonNode
...
-	}
+		{ "low", CBuilding::HEIGHT_LOW },
+		{ "average", CBuilding::HEIGHT_AVERAGE },
+		{ "high", CBuilding::HEIGHT_HIGH },
+		{ "skyship", CBuilding::HEIGHT_SKYSHIP }
+	};
+
+	static const std::map<std::string, BuildingSubID::EBuildingSubID> SPECIAL_BUILDINGS =
+	{
+		{ "mystic_pond", BuildingSubID::MYSTIC_POND },
```
Comment (alexvins) [inline lib/CTownHandler.cpp:394]: Please use lowerCamelCase for JSON keys.

**Review by alexvins — CHANGES_REQUESTED** [review]

File: lib/mapObjects/CGTownInstance.cpp [hunk 6]
```diff
@@ -1136,15 +1158,35 @@ void CGTownInstance::recreateBuildingsBonuses()
...
+		for (const auto & it : town->buildings)
+		{
+			if (it.second->subId == building)
+			{
+				descr << it.second->Name();
+				break;
+			}
+		}
+		auto b = std::make_shared<Bonus>(Bonus::PERMANENT, type, Bonus::TOWN_STRUCTURE, val, building, descr.str(), subtype);
+		if (prop)
```
Comment (alexvins) [inline lib/mapObjects/CGTownInstance.cpp:1184]: `prop` is not set anywhere,

**Review by alexvins — CHANGES_REQUESTED** [review]

**Review by ShubusCorporation — COMMENTED** [review]
@alexvins ,  builds are ok 
https://travis-ci.org/github/vcmi/vcmi/builds/732379045

File: lib/CTownHandler.h [hunk 7]
```diff
@@ -78,6 +88,17 @@ class DLL_LINKAGE CBuilding
...
+
+		if(version >= 792)
+		{
+			h & subId;
+			h & height;
+		}
+		else if (!h.saving)
+		{
+			subId = BuildingSubID::NONE;
+			height = CBuilding::HEIGHT_NO_TOWER;
```
Comment (alexvins) [inline lib/CTownHandler.h:100]: We should set orrect `height` and `subId` for old saves.

File: lib/mapObjects/CGTownInstance.h [hunk 8]
```diff
@@ -97,25 +97,41 @@ class DLL_LINKAGE CGTownBuilding : public IObjectInterface
...
+		h & indexOnTV;
+
+		if(version >= 792)
+			h & bType;
+		else if(!h.saving)
+			bType = BuildingSubID::NONE;
```
Comment (alexvins) [inline lib/mapObjects/CGTownInstance.h:118]: We should set correct `bType` for old saves.

**Review by alexvins — CHANGES_REQUESTED** [review]

**Comment by ShubusCorporation** [comment]
It seems, at the moment of loading game from the old save, this information is already nowhere to take:

load_old_save

**Comment by ShubusCorporation** [comment]
Furthermore, in CTown the same approach is used and it does't carry about actual moat hexes:

`	template <typename Handler> void serialize(Handler &h, const int version)
	{
		h & names;
		h & faction;
		h & creatures;
		h & dwellings;
		h & dwellingNames;
		h & buildings;
		h & hordeLvl;
		h & mageLevel;
		h & primaryRes;
		h & warMachine;
		h & clientInfo;
		h & moatDamage;
		if(version >= 758)
		{
			h & moatHexes;
		}
		else if(!h.saving)
		{
			moatHexes = defaultMoatHexes();
		}
		h & defaultTavernChance;
	}`

**Comment by alexvins** [comment]
> It seems, at the moment of loading from the old save, this information is already nowhere to take:
> 
An example of how to solve this is `CSpellHandler::update780`

**Review by alexvins — APPROVED** [review]

**Comment by ShubusCorporation** [comment]
Any objections on merging this?

**Pull request merged**